src/java/fr/paris/lutece/portal/service/download/AbstractFileDownloadProvider.java

```diff
@@ -0,0 +1,135 @@
+        }
+        catch( GeneralSecurityException e )
+        {
+            AppLogService.error( e.getMessage( ), e );
+        }
+        return null;
+    }
+    
+    @Override
+    public final File getFile( HttpServletRequest request ) throws SiteMessageException
+    {
+        File file = null;
+        try
+        {
+            String data = request.getParameter( PARAM_DATA );
+            String decrypted = RsaService.decryptRsa( data  );
+            FileDownloadData fileDownloadData = getDecryptedData( decrypted );
+            checkLinkValidity( request, fileDownloadData );
+            
+            file = FileHome.findByPrimaryKey( fileDownloadData.getIdFile( ) );
+
+            if ( file != null && file.getPhysicalFile( ) != null )
+            {
+                file.setPhysicalFile( PhysicalFileHome.findByPrimaryKey( file.getPhysicalFile( ).getIdPhysicalFile( ) ) );
+            }
+        }
+        catch( GeneralSecurityException e )
+        {
+            AppLogService.error( e.getMessage( ), e );
+        }
+        catch( ExpiredLinkException e )
+        {
+           AppLogService.error( e.getMessage( ), e );
+           SiteMessageService.setMessage( request, MESSAGE_LINK_EXPIRED );
+        }
+        return file;
+    }
+    
+    /**
+     * Find provider by name
+     * @param providerName
+     * @return
+     */
+    public static IFileDownloadProvider findProvider( String providerName )
+    {
+        for ( IFileDownloadProvider p : SpringContextService.getBeansOfType( IFileDownloadProvider.class ) )
+        {
+            if ( providerName.equals( p.getProviderName( ) ) )
+            {
+                return p;
+            }
+        }
+        return null;
+    }
+    
+    private String getDataToEncrypt( FileDownloadData fileDownloadData )
+    {
+        StringBuilder sb = new StringBuilder( );
+        sb.append( fileDownloadData.getIdFile( ) ).append( SEPARATOR );
+        sb.append( fileDownloadData.getIdResource( ) ).append( SEPARATOR );
+        sb.append( fileDownloadData.getResourceType( ) ).append( SEPARATOR );
+        sb.append( calculateEndValidity( ) );
+        return sb.toString( );
+    }
+    
+    private FileDownloadData getDecryptedData( String data )
+    {
+        String[] dataArray = data.split( SEPARATOR );
+        FileDownloadData fileDownloadData = new FileDownloadData( Integer.parseInt( dataArray[1] ) , dataArray[2], Integer.parseInt( dataArray[0] ) );
+        fileDownloadData.setEndValidity( new Timestamp( Long.parseLong( dataArray[3] ) ).toLocalDateTime( ) );
+                
+        return fileDownloadData;
+    }
+    
+    private void checkLinkValidity( HttpServletRequest request, FileDownloadData fileDownloadData ) throws ExpiredLinkException
+    {
+        if ( LocalDateTime.now( ).isAfter( fileDownloadData.getEndValidity( ) ) )
+        {
+            throw new ExpiredLinkException( "Link expired on : " + fileDownloadData.getEndValidity( ).toString( ) );
```

Comment:
Maybe the encrypted link data should be logged. If not, it is not possible to know which link is in error